source/extensions/common/http_wasm/context.h

```diff
@@ -0,0 +1,241 @@
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap& trailers) override;
+  Http::FilterMetadataStatus decodeMetadata(Http::MetadataMap& metadata_map) override;
+  void setDecoderFilterCallbacks(Envoy::Http::StreamDecoderFilterCallbacks& callbacks) override;
+
+  // Http::StreamEncoderFilter
+  Http::Filter1xxHeadersStatus encode1xxHeaders(Http::ResponseHeaderMap&) override;
+  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterDataStatus encodeData(::Envoy::Buffer::Instance& data, bool end_stream) override;
+  Http::FilterTrailersStatus encodeTrailers(Http::ResponseTrailerMap& trailers) override;
+  Http::FilterMetadataStatus encodeMetadata(Http::MetadataMap& metadata_map) override;
+  void setEncoderFilterCallbacks(Envoy::Http::StreamEncoderFilterCallbacks& callbacks) override;
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers, bool end_of_stream);
+  FilterDataStatus onRequestBody();
+  FilterTrailersStatus onRequestTrailers(uint32_t trailers);
+  FilterMetadataStatus onRequestMetadata(uint32_t elements);
+  FilterHeadersStatus onResponseHeaders(uint32_t headers, bool end_of_stream);
+  FilterDataStatus onResponseBody(uint32_t body_length, bool end_of_stream);
+  FilterTrailersStatus onResponseTrailers(uint32_t trailers);
+  FilterMetadataStatus onResponseMetadata(uint32_t elements);
+
+  bool isFailed();
+
+  // General
+  WasmResult log(int32_t level, std::string_view message);
+  std::string_view getConfiguration();
+  void sendLocalResponse(WasmBufferType);
+  void setLocalResponseCode(uint32_t);
+  void setLocalResponseBody(std::string_view);
+  void clearLocalResponse() {
+    local_response_body_ = "";
+    local_response_status_code_ = 0;
+    local_response_headers_.reset();
+  }
+
+  // Header/Trailer/Metadata Maps
+  WasmResult addHeaderMapValue(WasmHeaderMapType type, std::string_view key,
+                               std::string_view value);
+  WasmResult getHeaderMapValue(WasmHeaderMapType type, std::string_view key,
+                               std::vector<std::string_view>& name_values);
+  WasmResult getHeaderNames(WasmHeaderMapType type, std::vector<std::string_view>& names);
+  WasmResult removeHeaderMapValue(WasmHeaderMapType type, std::string_view key);
+  WasmResult replaceHeaderMapValue(WasmHeaderMapType type, std::string_view key,
+                                   std::string_view value);
+
+  WasmResult getHeaderMapSize(WasmHeaderMapType type, uint32_t* size);
+
+  // Buffer
+  WasmBuffer* getBuffer(WasmBufferType type);
+
+  uint64_t getCurrentTimeNanoseconds() {
+    error("unimplemented wasi API");
+    return 0;
+  }
+  uint64_t getMonotonicTimeNanoseconds() {
+    error("unimplemented wasi API");
+    return 0;
+  }
+
+  void setBufferRequest() { guest_feature_set_ = guest_feature_set_ | 1; }
+  void setBufferResponse() { guest_feature_set_ = guest_feature_set_ | 2; }
```

Comment:
Let's have excplicit names
```suggestion
  void setFeatureBufferRequest() { guest_feature_set_ = guest_feature_set_ | 1; }
  void setFeatureBufferResponse() { guest_feature_set_ = guest_feature_set_ | 2; }
```

Reply:
and can we use the bit rep?
```suggestion
  void setBufferRequest() { guest_feature_set_ = guest_feature_set_ | 0b1; }
  void setBufferResponse() { guest_feature_set_ = guest_feature_set_ | 0b10; }
```